### PR TITLE
Expose the promote mode

### PR DESCRIPTION
**Pull request by diml**
This PR exposes the promote mode that was internally implemented in #370.

## `(mode ...)` field

This PR adds a `(mode <mode>)` field to `rule` and `menhir` stanzas. The following modes are allowed:

- `standard`, the default
- `fallback`, replaces `(fallback)`
- `promote`, after the rule has been executed copy the targets to the source tree
- `promote-until-clean`, same as `promote` except that `jbuilder clean` delete the files copied to the source tree

## `ocamlyacc` and `ocamllex` extensions

To allow using a different mode with `ocamlyacc` and `ocamllex` stanzas, an alternative syntax is added:

```scheme
(ocamlyacc ((modules (<modules>)) (mode <mode)))
(ocamllex ((modules (<modules>)) (mode <mode)))
```

## `--ignore-promoted-rules` argument

The `--ignore-promoted-rules` argument is added to make jbuilder filter out rules with `(mode promote)`. It is automatically set by `-p/--for-release-of-packages`, to cut down dependencies during releases.

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -71,6 +71,18 @@ next
 - Use /Fo instead of -o when invoking the Microsoft C compiler to eliminate
   deprecation warning when compiling C++ sources (#354)
 
+- Add a mode field to `rule` stanzas:
+  + `(mode standard)` is the default
+  + `(mode fallback)` replaces `(fallback)`
+  + `(mode promote)` means that targets are copied to the source tree
+  after the rule has completed
+  + `(mode promote-until-clea)` is the same as `(mode promote)` except
```
Comment (samoht) [inline CHANGES.md]: `clea` -> `clean`

Comment (diml) [reply CHANGES.md]: indeed, thanks

File: src/jbuild.ml [hunk 2]
```diff
@@ -762,19 +772,51 @@ module Rule = struct
...
-         field_b "fallback" >>= fun fallback ->
+         map_validate
+           (field_b "fallback" >>= fun fallback ->
+            field_o "mode" Mode.t >>= fun mode ->
+            return (fallback, mode))
+           ~f:(function
+             | true, Some _ ->
+               Error "Cannot use both (fallback) and (mode ...) at the same time.\n\
+                      (fallback) is the same as (mode fallback), \
+                      please use the latter in new code."
```
Comment (rgrinberg) [inline src/jbuild.ml]: Will `(fallback)` be removed in 1.0?

Comment (diml) [reply src/jbuild.ml]: yes, we'll get the upgrade tool to translate it to `(mode fallback)`

**Pull request merged**